### PR TITLE
Support for `buildArgs` on Docker builds

### PR DESCRIPTION
**Pull request by aaronpowell**
Fixing #1270

Adding support for passing `buildArgs` to docker via the azure.yaml file

### CODE REVIEW

**Comment by weikanglim** [comment]
@aaronpowell Changes looks fine, but I'll probably want to hold this for team input to discuss on #1270 about other considerations of the feature.

File: cli/azd/pkg/project/framework_service_docker.go
Comment (wbreza) [inline cli/azd/pkg/project/framework_service_docker.go:26]: Lets add support for 1 or many build args.  Based on my research it looks like each docker build arg should be set as a separate argument.

Example
```
docker build \
-t essearch/ess-elasticsearch:1.7.6 \
--build-arg number_of_shards=5 \
--build-arg number_of_replicas=2 \
--no-cache .
```

In this case the `BuildArgs` property here should be updated to a `[]string` and then we can do the right thing when calling the docker CLI.

Comment (wbreza) [inline cli/azd/pkg/project/framework_service_docker.go:26]:
```suggestion
	BuildArgs []string           `json:"buildArgs"`
```

File: cli/azd/pkg/tools/docker/docker.go
Comment (wbreza) [inline cli/azd/pkg/tools/docker/docker.go:71]:
```suggestion
	buildArgs []string,
```

File: cli/azd/pkg/tools/docker/docker.go
Comment (wbreza) [inline cli/azd/pkg/tools/docker/docker.go:89]: Then here if buildArgs is defined we can go ahead and add an additional arg for each item

**Review by wbreza — CHANGES_REQUESTED** [review]
This is a great contribution, thank you! Can we update this to support multiple build args?

**References**
- https://docs.docker.com/engine/reference/commandline/build/#build-arg
- https://stackoverflow.com/questions/42297387/docker-build-with-build-arg-with-multiple-arguments

**Review by ellismg — APPROVED** [review]
LGTM - one question about masking build args in logs (fine for us to do in a follow up, if we aren't sure what the right thing is now).

I'll let @wbreza sign off and merge, but it looks good to me.

Thanks for this, @aaronpowell!

File: cli/azd/pkg/project/framework_service_docker.go
Comment (aaronpowell) [reply cli/azd/pkg/project/framework_service_docker.go:132]: @ellismg that's a really good point. A common use-case I see would be passing in an API key so that there can be some data fetched during the build process or accessing an authenticated package (npm, nuget, etc.) feed, so you don't want that in the logs.

I'll add it to this PR as I'd prefer we ship it right the first time.

Comment (aaronpowell) [reply cli/azd/pkg/project/framework_service_docker.go:138]: @ellismg I've implemented this, but I'm not sure it's the most efficient way to do it, since the existing redaction code was tied to how we execute commands, not how we write to `log`, so I had to adapt it.

**Review by savannahostrowski — APPROVED** [review]

**Review by wbreza — APPROVED** [review]
Thanks for making the requested changes.  This version looks great!

File: cli/azd/pkg/project/framework_service_docker.go
Comment (ellismg) [reply cli/azd/pkg/project/framework_service_docker.go:138]: Thanks, @aaronpowell, this looks like a good approach given the setup we had!

**Pull request merged**
